### PR TITLE
Script to install assisted-installer's operator

### PR DESCRIPTION
**Pull request by flaper87**
An operator has been created and released for the assisted-installer, which simplifies the deployment process for it and allows us to easily run it as part of a dev-script deployment without other node requirements.

This script uses the assisted-service operator to create an AI deployment. A few things are required:

- Local Storage operator (because AI requires 2 PVs)
- Hive operator (so that the ClusterDeployment CRDs are installed)

Missing (will do in separate PR)

- [ ] Validate that `VM_EXTRADISKS` was enabled and that nodes have the right number of disks

Signed-off-by: Flavio Percoco <flavio@redhat.com>

### CODE REVIEW

File: assisted_deployment.sh [hunk 1]
```diff
@@ -7,18 +7,181 @@ source common.sh
...
+  targetNamespaces:
+    - openshift-local-storage
+---
+apiVersion: operators.coreos.com/v1alpha1
+kind: Subscription
+metadata:
+  name: local-storage-operator
+  namespace: openshift-local-storage
+spec:
+  channel: "4.7"
```
Comment (flaper87) [inline assisted_deployment.sh:36]: I have to try without setting the `channel`. It doesn't sound right to hardcode this value. I tried using `$OPENSHIFT_VERSION` but it didn't seem to work. Will re-test.

File: assisted_deployment.sh [hunk 2]
```diff
@@ -7,18 +7,181 @@ source common.sh
 source assisted_deployment_utils.sh
 
 
+ASSISTED_INSTALL_TEST_INFRA=${ASSISTED_INSTALL_TEST_INFRA:-}
+ASSISTED_INSTALLER_NAMESPACE="${ASSISTED_INSTALLER_NAMESPACE:-assisted-installer}"
+ASSISTED_OPERATOR_INDEX="${ASSISTED_OPERATOR_INDEX:-quay.io/ocpmetal/assisted-service-index:latest}"
```
Comment (celebdor) [inline assisted_deployment.sh:12]: Should we have also an override for OpenShift Versions to put in the subscriptions?

File: assisted_deployment.sh [hunk 3]
```diff
@@ -7,18 +7,181 @@ source common.sh
...
+    - devicePaths:
+        - /dev/sdb
+        - /dev/sdc
+      storageClassName: assisted-service
+      volumeMode: Filesystem
+EOCR
+}
+
+
+function deploy_hive() {
```
Comment (celebdor) [inline assisted_deployment.sh:65]: does this obsolete hive_install.sh?

File: assisted_deployment.sh (hunk 2 above)
Comment (flaper87) [reply assisted_deployment.sh:12]: Do you mean as the target openshift version for AI? Or the `OPENSHIFT_VERSIONS` var?

Prob worth it, yeah.

File: assisted_deployment.sh (hunk 3 above)
Comment (flaper87) [reply assisted_deployment.sh:65]: Likely, and prob also worth removing the test-infra based install below.

I wanted to leave any removals/depreciations for a follow up PR as I'm not sure what may be using this code.

File: assisted_deployment.sh (hunk 2 above)
Comment (celebdor) [reply assisted_deployment.sh:12]: The OPENSHIFT_VERSIONS var, since I see people tweaking it all the time

File: assisted_deployment.sh [hunk 4]
```diff
@@ -7,18 +7,181 @@ source common.sh
...
+metadata:
+  name: assisted-service-operator
+  namespace: $ASSISTED_INSTALLER_NAMESPACE
+spec:
+  channel: alpha
+  installPlanApproval: Automatic
+  name: assisted-service-operator
+  source: assisted-service-catalog
+  sourceNamespace: openshift-marketplace
+  startingCSV: assisted-service-operator.v0.0.1
```
Comment (djzager) [inline assisted_deployment.sh:133]: For dev-scripts sake, we might as well make all the configuration possible...possible.

```suggestion
  startingCSV: assisted-service-operator.v0.0.1
    spec:
    channel: alpha
    config:
      env:
      - name: OPENSHIFT_VERSIONS
        value: ${OPENSHIFT_VERSIONS}
      - name: SERVICE_IMAGE
        value: ${ASSISTED_INSTALLER_SERVICE_IMAGE}
      - name: DATABASE_IMAGE
        value: ${ASSISTED_INSTALLER_DB_IMAGE}
      - name: AGENT_IMAGE
        value: ${ASSISTED_INSTALLER_AGENT_IMAGE}
      - name: CONTROLLER_IMAGE
        value: ${ASSISTED_INSTALLER_CONTROLLER_IMAGE}
      - name: INSTALLER_IMAGE
        value: ${ASSISTED_INSTALLER_IMAGE}
```

File: assisted_deployment.sh [hunk 5]
```diff
@@ -7,18 +7,232 @@ source common.sh
 source assisted_deployment_utils.sh
 
 
+ASSISTED_SERVICE_SERVICE_IMAGE="${ASSISTED_SERVICE_SERVICE_IMAGE:-}"
```
Comment (flaper87) [inline assisted_deployment.sh:10]: Will document before removing WIP :)

Comment (djzager) [reply assisted_deployment.sh:10]: Maybe it's enough to just make it `ASSISTED_*`?

File: assisted_deployment.sh [hunk 6]
```diff
@@ -7,18 +7,232 @@ source common.sh
...
+  cat <<EOF | oc apply -f -
+apiVersion: operators.coreos.com/v1alpha1
+kind: Subscription
+metadata:
+  name: assisted-service-operator
+  namespace: $ASSISTED_SERVICE_NAMESPACE
+spec:
+  config:
+    env:
+$(subscription_config)
```
Comment (djzager) [inline assisted_deployment.sh:179]: only curious what OLM does when nothing is specified here.

**Review by djzager — COMMENTED** [review]
Couple questions. Looking good. Thanks for working on this.

File: assisted_deployment.sh (hunk 6 above)
Comment (flaper87) [reply assisted_deployment.sh:180]: I acutally tested this. This results in `conffig: {}` and things still deploy normally 

DeepinScreenshot_select-area_20210419101938

DeepinScreenshot_select-area_20210419101953

File: assisted_deployment.sh [hunk 7]
```diff
@@ -4,28 +4,241 @@ set -o pipefail
...
+kind: LocalVolume
+metadata:
+  name: assisted-service
+  namespace: openshift-local-storage
+spec:
+  logLevel: Normal
+  managementState: Managed
+  storageClassDevices:
+    - devicePaths:
+        - /dev/sdb
```
Comment (rwsu) [inline assisted_deployment.sh:64]: Where are the extra disks being created for the LocalVolume?

Comment (flaper87) [reply assisted_deployment.sh:64]: Right now, manually! 

dev-scripts has an option to set `VM_EXTRADISKS` to true which will create one disk. This is not enough, so @lranjbar will work on a PR to allow dev-scripts for creating a custom number of disks that we can then use here.

This comment does raise a good point. Right now these disks are not configurable and it requires modifying this file manually. Wonder if it's worth making it configurable in this PR or just wait for @lranjbar PR

File: docs/assisted-deployment.md [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+Pre-requisites
+==
+
+2 disks need to be created and added to the worker nodes. This can be done either using
+`virt-manager` or libvirt directly. The disks should be of, at least, 10GB each and they should be
+attached to the `/dev/sdb` and `/dev/sdc` devices respectively. Please update
+`assisted_deployment.sh` if this is not the case.
+
+Note that there is work in progress to automate the disk creations and the device name
```
Comment (djzager) [inline docs/assisted-deployment.md:16]: Might be appropriate to create an issue and link it here.

File: config_example.sh [hunk 9]
```diff
@@ -219,7 +219,33 @@ set -x
...
+# export ASSISTED_DATABASE_IMAGE="quay.io/ocpmetal/postgresql-12-centos7:latest"
+# export ASSISTED_CONTROLLER_IMAGE="quay.io/ocpmetal/assisted-installer-controller:latest"
+
+# Get the latest default versions from the assisted-service repo itself.
+# export ASSISTED_OPENSHIFT_VERSIONS=$(wget -qO- https://raw.githubusercontent.com/openshift/assisted-service/master/default_ocp_versions.json)
+
+# Operator's bundle index to use. This will allow for testing a custom Assisted operator build that has
+# not been published yet. For custom Assisted Service images it's enough to overwrite one of the images above.
+# Overwrite this image only if you are working on the operator itself.
+# export ASSISTED_OPERATOR_INDEX="${quay.io/ocpmetal/assisted-service-index:latest}"
```
Comment (djzager) [inline config_example.sh:246]:
```suggestion
# export ASSISTED_OPERATOR_INDEX="quay.io/ocpmetal/assisted-service-index:latest"
```

**Review by djzager — COMMENTED** [review]
A couple little comments. Looks good to me.

File: config_example.sh (hunk 9 above)
Comment (flaper87) [reply config_example.sh:246]: whoops, thanks!

File: docs/assisted-deployment.md (hunk 8 above)
Comment (flaper87) [reply docs/assisted-deployment.md:16]: done :)

**Review by djzager — APPROVED** [review]

**Comment by stbenjam** [comment]
/approve
/lgtm
/override ci/prow/e2e-metal-ipi-ovn-ipv6 

None of this code is used by CI, and failure looks like a manifestation of https://bugzilla.redhat.com/show_bug.cgi?id=1948311.

BTW, @djzager I invited you to the openshift-metal3 org so you can /lgtm PR's.

**Comment by djzager** [comment]
> BTW, @djzager I invited you to the openshift-metal3 org so you can /lgtm PR's.

Alt Text

**Pull request merged**